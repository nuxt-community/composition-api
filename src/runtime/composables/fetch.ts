import {
  isRef,
  isReactive,
  onBeforeMount,
  onServerPrefetch,
  reactive,
  set,
} from '@vue/composition-api'

import {
  globalContext,
  globalNuxt,
  isFullStatic,
} from '@nuxtjs/composition-api/dist/runtime/globals'
import type { NuxtApp } from '@nuxt/types/app'

import { getCurrentInstance, ComponentInstance } from './utils'

const nuxtState = process.client && (window as any)[globalContext]

function normalizeError(err: any) {
  let message: string
  if (!(err.message || typeof err === 'string')) {
    try {
      message = JSON.stringify(err, null, 2)
    } catch (e) {
      message = `[${err.constructor.name}]`
    }
  } else {
    message = err.message || err
  }
  return {
    ...err,
    message,
    statusCode:
      err.statusCode ||
      err.status ||
      (err.response && err.response.status) ||
      500,
  }
}

function createGetCounter(counterObject: Record<string, any>, defaultKey = '') {
  return function getCounter(id = defaultKey) {
    if (counterObject[id] === undefined) {
      counterObject[id] = 0
    }
    return counterObject[id]++
  }
}

interface Fetch {
  (context: ComponentInstance): void | Promise<void>
}

const fetches = new WeakMap<ComponentInstance, Fetch[]>()
const fetchPromises = new Map<Fetch, Promise<any>>()

const isSsrHydration = (vm: ComponentInstance) =>
  (vm.$vnode?.elm as any)?.dataset?.fetchKey

interface AugmentedComponentInstance extends ComponentInstance {
  _fetchKey?: number | string
  _data?: any
  _hydrated?: boolean
  _fetchDelay?: number
  _fetchOnServer?: boolean
}

interface AugmentedNuxtApp extends NuxtApp {
  isPreview?: boolean
  _payloadFetchIndex?: number
  _pagePayload?: any
}

function registerCallback(vm: ComponentInstance, callback: Fetch) {
  const callbacks = fetches.get(vm) || []
  fetches.set(vm, [...callbacks, callback])
}

async function callFetches(this: AugmentedComponentInstance) {
  const fetchesToCall = fetches.get(this)
  if (!fetchesToCall) return
  ;(this[globalNuxt] as any).nbFetching++

  this.$fetchState.pending = true
  this.$fetchState.error = null
  this._hydrated = false

  let error = null
  const startTime = Date.now()

  try {
    await Promise.all(
      fetchesToCall.map(fetch => {
        if (fetchPromises.has(fetch)) return fetchPromises.get(fetch)
        const promise = Promise.resolve(fetch(this)).finally(() =>
          fetchPromises.delete(fetch)
        )
        fetchPromises.set(fetch, promise)
        return promise
      })
    )
  } catch (err) {
    if ((process as any).dev) {
      console.error('Error in fetch():', err)
    }
    error = normalizeError(err)
  }

  const delayLeft = (this._fetchDelay || 0) - (Date.now() - startTime)
  if (delayLeft > 0) {
    await new Promise(resolve => setTimeout(resolve, delayLeft))
  }

  this.$fetchState.error = error
  this.$fetchState.pending = false
  this.$fetchState.timestamp = Date.now()

  this.$nextTick(() => (this[globalNuxt] as any).nbFetching--)
}

const setFetchState = (vm: AugmentedComponentInstance) => {
  vm.$fetchState =
    vm.$fetchState ||
    reactive({
      error: null,
      pending: false,
      timestamp: 0,
    })
}

const mergeDataOnMount = (data: Record<string, any>) => {
  const vm = getCurrentInstance() as AugmentedComponentInstance | undefined
  if (!vm) throw new Error('This must be called within a setup function.')

  onBeforeMount(() => {
    // Merge data
    for (const key in data) {
      try {
        // Assign missing properties
        if (key in vm) {
          const _key = key as keyof typeof vm
          // Skip functions (not stringifiable)
          if (typeof vm[_key] === 'function') continue
          // Preserve reactive objects
          if (isReactive(vm[_key])) {
            // Unset keys that do not exist in incoming data
            for (const k in vm[_key]) {
              if (!(k in data[key])) {
                delete vm[_key][k]
              }
            }
            Object.assign(vm[_key], data[key])
            continue
          }
        }
        set(vm, key, data[key])
      } catch (e) {
        if (process.env.NODE_ENV === 'development')
          // eslint-disable-next-line
          console.warn(`Could not hydrate ${key}.`)
      }
    }
  })
}

const loadFullStatic = (vm: AugmentedComponentInstance) => {
  vm._fetchKey = getKey(vm)
  // Check if component has been fetched on server
  const { fetchOnServer } = vm.$options
  const fetchedOnServer =
    typeof fetchOnServer === 'function'
      ? fetchOnServer.call(vm) !== false
      : fetchOnServer !== false

  const nuxt = vm[globalNuxt] as AugmentedNuxtApp
  if (!fetchedOnServer || nuxt?.isPreview || !nuxt?._pagePayload) {
    return
  }
  vm._hydrated = true
  const data = nuxt._pagePayload.fetch[vm._fetchKey!]

  // If fetch error
  if (data && data._error) {
    vm.$fetchState.error = data._error
    return
  }

  mergeDataOnMount(data)
}

async function serverPrefetch(vm: AugmentedComponentInstance) {
  if (!vm._fetchOnServer) return

  // Call and await on $fetch
  setFetchState(vm)

  try {
    await callFetches.call(vm)
  } catch (err) {
    if ((process as any).dev) {
      console.error('Error in fetch():', err)
    }
    vm.$fetchState.error = normalizeError(err)
  }
  vm.$fetchState.pending = false

  // Define an ssrKey for hydration
  vm._fetchKey =
    // Nuxt 2.15+ uses a different format - an object rather than an array
    'push' in vm.$ssrContext.nuxt.fetch
      ? vm.$ssrContext.nuxt.fetch.length
      : vm._fetchKey || vm.$ssrContext.fetchCounters['']++

  // Add data-fetch-key on parent element of Component
  if (!vm.$vnode.data) vm.$vnode.data = {}
  const attrs = (vm.$vnode.data.attrs = vm.$vnode.data.attrs || {})
  attrs['data-fetch-key'] = vm._fetchKey

  const data = { ...vm._data }
  Object.entries((vm as any).__composition_api_state__.rawBindings).forEach(
    ([key, val]) => {
      if (val instanceof Function || val instanceof Promise) return

      data[key] = isRef(val) ? val.value : val
    }
  )

  // Add to ssrContext for window.__NUXT__.fetch
  const content = vm.$fetchState.error
    ? { _error: vm.$fetchState.error }
    : JSON.parse(JSON.stringify(data))
  if ('push' in vm.$ssrContext.nuxt.fetch) {
    vm.$ssrContext.nuxt.fetch.push(content)
  } else {
    vm.$ssrContext.nuxt.fetch[vm._fetchKey!] = content
  }
}

function getKey(vm: AugmentedComponentInstance) {
  const nuxtState = vm[globalNuxt] as any
  if (process.server && 'push' in vm.$ssrContext.nuxt.fetch) {
    return undefined
  } else if (process.client && '_payloadFetchIndex' in nuxtState) {
    nuxtState._payloadFetchIndex = nuxtState._payloadFetchIndex || 0
    return nuxtState._payloadFetchIndex++
  }
  const defaultKey = (vm.$options as any)._scopeId || vm.$options.name || ''
  const getCounter = createGetCounter(
    process.server
      ? vm.$ssrContext.fetchCounters
      : (vm[globalNuxt] as any)._fetchCounters,
    defaultKey
  )

  const options: {
    fetchKey:
      | ((getCounter: ReturnType<typeof createGetCounter>) => string)
      | string
  } = vm.$options as any

  if (typeof options.fetchKey === 'function') {
    return options.fetchKey.call(vm, getCounter)
  } else {
    const key =
      'string' === typeof options.fetchKey ? options.fetchKey : defaultKey
    return key ? key + ':' + getCounter(key) : String(getCounter(key))
  }
}

/**
 * Versions of Nuxt newer than v2.12 support a [custom hook called `fetch`](https://nuxtjs.org/api/pages-fetch/) that allows server-side and client-side asynchronous data-fetching.

 * @param callback The async function you want to run.
 * @example

  ```ts
  import { defineComponent, ref, useFetch } from '@nuxtjs/composition-api'
  import axios from 'axios'

  export default defineComponent({
    setup() {
      const name = ref('')

      const { fetch, fetchState } = useFetch(async () => {
        name.value = await axios.get('https://myapi.com/name')
      })

      // Manually trigger a refetch
      fetch()

      // Access fetch error, pending and timestamp
      fetchState

      return { name }
    },
  })
  ```
 */
export const useFetch = (callback: Fetch) => {
  const vm = getCurrentInstance() as AugmentedComponentInstance | undefined
  if (!vm) throw new Error('This must be called within a setup function.')

  registerCallback(vm, callback)

  if (typeof vm.$options.fetchOnServer === 'function') {
    vm._fetchOnServer = vm.$options.fetchOnServer.call(vm) !== false
  } else {
    vm._fetchOnServer = vm.$options.fetchOnServer !== false
  }

  if (process.server) {
    vm._fetchKey = getKey(vm)
  }

  setFetchState(vm)

  onServerPrefetch(() => serverPrefetch(vm))

  function result() {
    return {
      fetch: vm!.$fetch,
      fetchState: vm!.$fetchState,
      $fetch: vm!.$fetch,
      $fetchState: vm!.$fetchState,
    }
  }

  vm._fetchDelay =
    typeof vm.$options.fetchDelay === 'number' ? vm.$options.fetchDelay : 0

  vm.$fetch = callFetches.bind(vm)

  onBeforeMount(() => !vm._hydrated && callFetches.call(vm))

  if (process.server || !isSsrHydration(vm)) {
    if (process.client && isFullStatic) loadFullStatic(vm)
    return result()
  }

  // Hydrate component
  vm._hydrated = true
  vm._fetchKey = (vm.$vnode.elm as any)?.dataset.fetchKey || getKey(vm)
  const data = nuxtState.fetch[vm._fetchKey!]

  // If fetch error
  if (data && data._error) {
    vm.$fetchState.error = data._error
    return result()
  }

  mergeDataOnMount(data)

  return result()
}
