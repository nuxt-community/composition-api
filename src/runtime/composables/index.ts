export { useAsync } from './async'
export { defineComponent } from './component'
export { callWithContext, useContext, withContext } from './context'
export * from './defineHelpers'
export { useFetch } from './fetch'
export { globalPlugin, onGlobalSetup, setMetaPlugin } from './hooks'
export { useMeta } from './meta'
export { reqRef, reqSsrRef } from './req-ref'
export { ssrRef, shallowSsrRef, setSSRContext, ssrPromise } from './ssr-ref'
export { useStatic } from './static'
export {
  useRoute,
  useRouter,
  useStore,
  wrapProperty,
  wrapContextProperty,
  useRedirect,
} from './wrappers'

export * from './vue'
