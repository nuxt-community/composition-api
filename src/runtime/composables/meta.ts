import defu from 'defu'
import {
  computed,
  reactive,
  ref,
  toRefs,
  watch,
  Ref,
  isReactive,
  isRef,
  toRaw,
  UnwrapRef,
} from 'vue'

import type { MetaInfo } from 'vue-meta'

import { getCurrentInstance } from './utils'

export type ReactiveHead<T = Record<string, unknown>> = UnwrapRef<
  Ref<MetaInfo & T>
>

type MetaInfoMapper<T> = {
  [P in keyof T]: P extends 'base'
    ? T[P] | undefined
    : T[P] extends () => any
    ? T[P] | undefined
    : T[P] extends Array<any> | Record<string, unknown>
    ? T[P]
    : T[P] | undefined
}

export function createEmptyMeta(): MetaInfoMapper<Required<MetaInfo>> {
  return {
    titleTemplate: null as unknown as undefined,

    __dangerouslyDisableSanitizers: [],
    __dangerouslyDisableSanitizersByTagID: {},

    title: undefined,
    htmlAttrs: {},
    headAttrs: {},
    bodyAttrs: {},

    base: undefined,

    meta: [],
    link: [],
    style: [],
    script: [],
    noscript: [],

    changed: undefined,
    afterNavigation: undefined,
  }
}

type ComputedHead = Array<MetaInfo | Ref<MetaInfo>>

export const getHeadOptions = (options: { head: () => MetaInfo }) => {
  const head = function (this: {
    head?: MetaInfo | (() => MetaInfo)
    _computedHead?: ComputedHead
  }) {
    const optionHead =
      options.head instanceof Function ? options.head.call(this) : options.head

    if (!this._computedHead) return optionHead

    const computedHead = this._computedHead.map(h => {
      if (isReactive(h)) return toRaw(h)
      if (isRef(h)) return h.value
      return h
    })
    return defu({} as MetaInfo, ...computedHead.reverse(), optionHead)
  }

  return { head }
}

type ToRefs<T extends Record<string, any>> = {
  [P in keyof T]: Ref<T[P]>
}

/**
 * `useMeta` lets you interact directly with [`head()` properties](https://nuxtjs.org/api/pages-head/) in `setup`. **Make sure you set `head: {}` in your component options.**
 * @example
    ```ts
    import { defineComponent, useMeta, computed } from '@nuxtjs/composition-api'

    export default defineComponent({
      head: {},
      setup() {
        const { title } = useMeta()
        title.value = 'My page'
      })
    })
    ```
 * @param init Whatever defaults you want to set for `head` properties.
 */
export const useMeta = <
  T extends MetaInfo,
  MetaRefs extends ToRefs<ReturnType<typeof createEmptyMeta> & T>
>(
  init?: T | (() => T)
) => {
  const vm = getCurrentInstance() as ReturnType<typeof getCurrentInstance> & {
    _computedHead?: ComputedHead
    _metaRefs?: MetaRefs
  }
  if (!vm) throw new Error('useMeta must be called within a component.')

  if (!('head' in vm.$options))
    throw new Error(
      'In order to enable `useMeta`, please make sure you include `head: {}` within your component definition, and you are using the `defineComponent` exported from @nuxtjs/composition-api.'
    )

  const refreshMeta = () => vm.$meta().refresh()

  if (!vm._computedHead) {
    const metaRefs = reactive(createEmptyMeta())
    vm._computedHead = [metaRefs]
    vm._metaRefs = toRefs(metaRefs) as MetaRefs

    if (process.client) {
      watch(Object.values(vm._metaRefs), refreshMeta, { immediate: true })
    }
  }

  if (init) {
    const initRef = init instanceof Function ? computed(init) : ref(init)
    vm._computedHead.push(initRef)

    if (process.client) {
      watch(initRef, refreshMeta, { immediate: true })
    }
  }

  return vm._metaRefs!
}
