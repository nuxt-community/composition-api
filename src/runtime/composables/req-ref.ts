import { ref, Ref } from 'vue'

import { sanitise, ssrRef } from './ssr-ref'

export const reqRefs = new Set<() => void>()

/**
 * @deprecated
 */
export const reqRef = <T>(initialValue: T): Ref<T> => {
  const _ref = ref(initialValue)

  if (process.server) reqRefs.add(() => (_ref.value = initialValue as any))

  return _ref as Ref<T>
}

/**
 * @deprecated
 */
export const reqSsrRef = <T>(initialValue: T, key?: string) => {
  const _ref = ssrRef(initialValue, key)

  if (process.server)
    reqRefs.add(() => {
      _ref.value =
        initialValue instanceof Function
          ? sanitise(initialValue())
          : initialValue
    })

  return _ref
}
