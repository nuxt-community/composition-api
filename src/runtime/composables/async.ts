import { isRef, onServerPrefetch } from 'vue'
import type { Ref } from 'vue'

import { globalNuxt } from '@nuxtjs/composition-api/dist/runtime/globals'
import { ssrRef } from './ssr-ref'
import { validateKey } from './utils'

/**
 * You can create reactive values that depend on asynchronous calls with `useAsync`.
 * On the server, this helper will inline the result of the async call in your HTML and automatically inject them into your client code. Much like `asyncData`, it _won't_ re-run these async calls client-side.
 *
 * However, if the call hasn't been carried out on SSR (such as if you have navigated to the page after initial load), it returns a `null` ref that is filled with the result of the async call when it resolves.
 *
 * **At the moment, `useAsync` is only suitable for one-offs, unless you provide your own unique key.**
 * @param cb The async function that will populate the ref this function returns.
 * @param key Under the hood, `useAsync` requires a key to ensure that the ref values match between client and server. If you have added `@nuxtjs/composition-api/module` to your `buildModules`, this will be done automagically by an injected Babel plugin. If you need to do things differently, you can specify a key manually or add `@nuxtjs/composition-api/dist/babel-plugin` to your Babel plugins.
 *
 * @example
  ```ts
  import { defineComponent, useAsync, computed } from '@nuxtjs/composition-api'
  import axios from 'axios'

  export default defineComponent({
    setup() {
      const posts = useAsync(() => axios.get('/api/posts'))

      return { posts }
    },
  })
  ```
 */
export const useAsync = <T>(
  cb: () => T | Promise<T>,
  key?: string | Ref<null>
): Ref<null | T> => {
  validateKey(key)

  const _ref = isRef(key) ? key : ssrRef<T | null>(null, key)

  if (
    !_ref.value ||
    (process.env.NODE_ENV === 'development' &&
      process.client &&
      window[globalNuxt]?.context.isHMR)
  ) {
    const p = Promise.resolve(cb())

    if (process.server) {
      onServerPrefetch(async () => {
        _ref.value = await p
      })
    } else {
      // eslint-disable-next-line
      p.then(res => (_ref.value = res))
    }
  }

  return _ref as Ref<null | T>
}
