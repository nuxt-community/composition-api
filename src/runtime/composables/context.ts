import { computed } from 'vue'

import type { Ref } from 'vue'
import type { Context } from '@nuxt/types'
import type { Route } from 'vue-router'

import { globalNuxt } from '@nuxtjs/composition-api/dist/runtime/globals'
import { getCurrentInstance } from './utils'

interface ContextCallback {
  (context: Context): void
}

/**
 * @deprecated
 * Recommend using `useContext` instead
 */
export const withContext = (callback: ContextCallback) => {
  const vm = getCurrentInstance()
  if (!vm) throw new Error('This must be called within a setup function.')

  callback((vm[globalNuxt] || vm.$options).context)
}

interface UseContextReturn
  extends Omit<Context, 'route' | 'query' | 'from' | 'params'> {
  route: Ref<Route>
  query: Ref<Route['query']>
  from: Ref<Context['from']>
  params: Ref<Route['params']>
}

/**
 * `useContext` will return the Nuxt context.
 * @example
  ```ts
  import { defineComponent, ref, useContext } from '@nuxtjs/composition-api'

  export default defineComponent({
    setup() {
      const { store } = useContext()
      store.dispatch('myAction')
    },
  })
  ```
 */
export const useContext = (): UseContextReturn => {
  const vm = getCurrentInstance()
  if (!vm) throw new Error('This must be called within a setup function.')

  return {
    ...(vm[globalNuxt] || vm.$options).context,
    /**
     * @deprecated To smooth your upgrade to Nuxt 3, it is recommended not to access `route` from `useContext` but rather to use the `useRoute` helper function.
     */
    route: computed(() => vm.$route),
    /**
     * @deprecated To smooth your upgrade to Nuxt 3, it is recommended not to access `query` from `useContext` but rather to use the `useRoute` helper function.
     */
    query: computed(() => vm.$route.query),
    /**
     * @deprecated To smooth your upgrade to Nuxt 3, it is recommended not to access `from` from `useContext` but rather to use the `useRoute` helper function.
     */
    from: computed(() => (vm[globalNuxt] || vm.$options).context.from),
    /**
     * @deprecated To smooth your upgrade to Nuxt 3, it is recommended not to access `params` from `useContext` but rather to use the `useRoute` helper function.
     */
    params: computed(() => vm.$route.params),
  }
}
