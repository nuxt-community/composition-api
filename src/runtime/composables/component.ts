import { defineComponent as define } from 'vue'

import { getHeadOptions } from './meta'

/**
 * If you want to enable `useMeta`, make sure to include `head: {}` in your component definition.
 * @example
  ```ts
  import { defineComponent } from '@nuxtjs/composition-api'

  export default defineComponent({
    head: {},
    setup() {
      ...
    }
  })
  ```
 */
export const defineComponent: typeof define = (options: any) => {
  if (!('head' in options)) return options

  return {
    ...options,
    ...getHeadOptions(options),
  }
}
