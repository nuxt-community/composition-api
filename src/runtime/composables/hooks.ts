import type { Plugin } from '@nuxt/types'
import type { SetupContext } from 'vue'
import { getHeadOptions } from './meta'

import { reqRefs } from './req-ref'
import { setSSRContext } from './ssr-ref'

type SetupFunction = (
  this: void,
  props: Record<string, unknown>,
  ctx: SetupContext
) => void | Record<any, any>

let globalSetup: Set<SetupFunction>

/**
 * Run a callback function in the global setup function. This should be called from a Nuxt plugin.
 * @param fn The function to run in the setup function. It receives the global props and context.
 * @example
    ```ts
    import { onGlobalSetup } from '@nuxtjs/composition-api'
    
    export default () => {
      onGlobalSetup(() => {
        provide('globalKey', true)
      })
    }
    ```
 */
export const onGlobalSetup = (fn: SetupFunction) => {
  globalSetup.add(fn)
}

/**
 *
 * @private
 */
export const setMetaPlugin: Plugin = context => {
  const { head } = context.app
  Object.assign(context.app, getHeadOptions({ head: head as any }))
}

/**
 * @private
 */
export const globalPlugin: Plugin = context => {
  if (process.server) {
    reqRefs.forEach(reset => reset())
    setSSRContext(context.app)
  }

  const { setup } = context.app
  globalSetup = new Set<SetupFunction>()

  context.app.setup = function (...args) {
    let result = {}
    if (setup instanceof Function) {
      result = setup(...args) || {}
    }
    for (const fn of globalSetup) {
      result = { ...result, ...(fn.call(this, ...args) || {}) }
    }
    return result
  }
}
