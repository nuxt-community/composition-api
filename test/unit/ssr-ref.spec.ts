/**
 * @jest-environment jsdom
 */
import { ssrRef, globalPlugin } from '../../src/runtime/composables'
import * as cAPI from '@vue/composition-api'

describe('ssrRef reactivity', () => {
  let ssrContext: Record<string, any>

  beforeEach(async () => {
    process.server = true
    ssrContext = Object.assign({}, { nuxt: {} })
    ;(cAPI as any).getCurrentInstance = () => ({
      proxy: {
        $nuxt: {
          context: { ssrContext },
        },
      },
    })
    globalPlugin({ app: { context: { ssrContext } } } as any, null)
  })
  test('ssrRefs react to change in state', async () => {
    process.client = false
    const name = ssrRef('', 'name')
    ssrRef('', 'unchanged')
    name.value = 'full name'
    expect(ssrContext).toMatchSnapshot()
  })
  test('ssrRefs react to deep change in object state', async () => {
    process.client = false
    const obj = ssrRef({ deep: { object: { name: 'nothing' } } }, 'obj')
    obj.value.deep.object.name = 'full name'
    expect(ssrContext).toMatchSnapshot()
  })
  test('ssrRefs react to deep change in array state', async () => {
    process.client = false
    const obj = ssrRef({ deep: { object: [{ name: 'nothing' }] } }, 'obj')
    obj.value.deep.object[0].name = 'full name'
    expect(ssrContext).toMatchSnapshot()
  })

  test('ssrRefs within multiple requests', async () => {
    const concurrentRef = ssrRef(1, 'concurrentRef')

    // simulate the new request comes in
    globalPlugin(
      { app: { context: { ssrContext: { nuxt: {} } } } } as any,
      null
    )

    concurrentRef.value = 2

    expect(ssrContext).toMatchSnapshot()
  })

  test('ssrRefs react to constructors', async () => {
    const testMap = new Map()
    testMap.set('john', 'doe')

    const obj = ssrRef({ testMap }, 'obj')

    expect(obj).toMatchSnapshot()
  })
})
