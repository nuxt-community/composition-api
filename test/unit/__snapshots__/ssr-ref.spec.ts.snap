// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ssrRef reactivity ssrRefs react to change in state 1`] = `
Object {
  "nuxt": Object {
    "globalRefs": Object {},
    "ssrRefs": Object {
      "name": "full name",
    },
  },
}
`;

exports[`ssrRef reactivity ssrRefs react to constructors 1`] = `
RefImpl {
  "value": bound Object {
    "testMap": Map {
      "john" => "doe",
    },
  },
}
`;

exports[`ssrRef reactivity ssrRefs react to deep change in array state 1`] = `
Object {
  "nuxt": Object {
    "globalRefs": Object {},
    "ssrRefs": Object {
      "obj": Object {
        "deep": Object {
          "object": Array [
            Object {
              "name": "full name",
            },
          ],
        },
      },
    },
  },
}
`;

exports[`ssrRef reactivity ssrRefs react to deep change in object state 1`] = `
Object {
  "nuxt": Object {
    "globalRefs": Object {},
    "ssrRefs": Object {
      "obj": Object {
        "deep": Object {
          "object": Object {
            "name": "full name",
          },
        },
      },
    },
  },
}
`;

exports[`ssrRef reactivity ssrRefs within multiple requests 1`] = `
Object {
  "nuxt": Object {
    "globalRefs": Object {},
    "ssrRefs": Object {
      "concurrentRef": 2,
    },
  },
}
`;
